core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -125,22 +129,30 @@ class TransactionCoordinator(brokerId: Int,
           val createdMetadata = new TransactionMetadata(transactionalId = transactionalId,
             producerId = producerId,
             producerEpoch = RecordBatch.NO_PRODUCER_EPOCH,
+            lastProducerEpoch = RecordBatch.NO_PRODUCER_EPOCH,
             txnTimeoutMs = transactionTimeoutMs,
             state = Empty,
             topicPartitions = collection.mutable.Set.empty[TopicPartition],
             txnLastUpdateTimestamp = time.milliseconds())
           txnManager.putTransactionStateIfNotExists(transactionalId, createdMetadata)
 
-        case Some(epochAndTxnMetadata) => Right(epochAndTxnMetadata)
+        case Some(epochAndTxnMetadata) =>
+          if (producerId != NO_PRODUCER_ID && epochAndTxnMetadata.transactionMetadata.producerId != producerId)
```

Comment:
Hmm... I think we might need to do this check under the `txnMetadata` lock below.